pkgs/leak_tracker/test/tests/leak_tracking/_object_tracker_test.dart

```diff
@@ -375,96 +389,197 @@ void main() {
     late _MockFinalizerBuilder finalizerBuilder;
     late _MockGcCounter gcCounter;
     late ObjectTracker tracker;
-    late ObjectRef<PhaseSettings> phase;
 
-    void startTracking(Object object) {
+    const objectsToPhases = <Object, PhaseSettings>{
+      '0': PhaseSettings.paused(),
+      '1': PhaseSettings(
+        name: '1',
+      ),
+      '2': PhaseSettings.paused(),
+      '3': PhaseSettings(
+        name: '3',
+      ),
+      '4': PhaseSettings(
+        name: '4',
+        leakDiagnosticConfig: LeakDiagnosticConfig(
+          collectRetainingPathForNotGCed: true,
+        ),
+      ),
+      '5': PhaseSettings(
+        name: '5',
+        leakDiagnosticConfig: LeakDiagnosticConfig(
+          collectStackTraceOnDisposal: true,
+        ),
+      ),
+      '6': PhaseSettings(
+        name: '6',
+        leakDiagnosticConfig: LeakDiagnosticConfig(
+          collectStackTraceOnStart: true,
+        ),
+      ),
+    };
+
+    void startTracking(Object object, PhaseSettings phase) {
       tracker.startTracking(
         object,
         trackedClass: _trackedClass,
         context: null,
+        phase: phase,
       );
     }
 
     setUp(() {
-      phase = ObjectRef(const PhaseSettings.paused());
       finalizerBuilder = _MockFinalizerBuilder();
       gcCounter = _MockGcCounter();
       tracker = ObjectTracker(
         finalizerBuilder: finalizerBuilder.build,
         gcCounter: gcCounter,
         disposalTime: _disposalTime,
         numberOfGcCycles: defaultNumberOfGcCycles,
-        maxRequestsForRetainingPath: 0,
-        phase: phase,
+        maxRequestsForRetainingPath: null,
         switches: const Switches(),
       );
     });
 
-    test('when object is registered before phase started.', () async {
-      const objectBeforePhase = '1';
-      const objectBeforePhaseLeaking = '2';
-      const objectInPhase = '3';
-      const objectInPhaseLeaking = '4';
-      final allObjects = [
-        objectBeforePhase,
-        objectBeforePhaseLeaking,
-        objectInPhase,
-        objectInPhaseLeaking,
-      ];
-
-      // Start tracking for all objects.
-      startTracking(objectBeforePhase);
-      startTracking(objectBeforePhaseLeaking);
-      // Start new phase
-      phase.value = const PhaseSettings();
-      startTracking(objectInPhase);
-      startTracking(objectInPhaseLeaking);
-
-      // Dispose non-leaking objects.
-      tracker.dispatchDisposal(objectInPhase, context: null);
-      tracker.dispatchDisposal(objectBeforePhase, context: null);
-
-      // GC all objects.
-      allObjects.forEach(finalizerBuilder.gc);
-
-      // Check leaks are collected only for leaking object that
-      // was registered in the phase.
-      final leaks = await tracker.collectLeaks();
-      expect(leaks.total, 1);
-      final theLeak = leaks.byType[LeakType.notDisposed]!.single;
-      expect(theLeak.code, identityHashCode(objectInPhaseLeaking));
-    });
+    void checkContext(
```

Comment:
to help readability of tests, it might be better to make these helper functions as custom matchers

Reply:
Interesting. I can think about it and maybe implement in other PRs.
Did you ever create custom matchers, not for publishing, but for local use, in your tests? If yes, can you give example?

Reply:
I have feeling that custom matcher implementation is more complicated than helper method, and create overhead for debugging the matcher itself, so it would be overdesign to do just for local use.

Reply:
I haven't written any myself yet, but it looks like it is easy enough: https://adevait.com/flutter/write-flutter-tests-custom-matchers

A couple years back I did a bunch of reading into unit test best practices, they mention that using helpers should be avoided since it makes tests harder to read. Something about readability and keeping the logic inside the main context.

All in all if these tests get the job done, then that is just a bit of a nit pick on my part

Reply:
Is this practice recommended by dart team? The article you linked is not.
The problem with matchers is that they are not strongly typed, so it is extra effort to do type promotion.